core/crypto/src/key_file.rs

```diff
@@ -22,10 +57,242 @@ impl KeyFile {
+            let key_hash = KeyHash::new(&password_input.content);
+            let ssb = SodiumoxideSecretBox::new();
+            let encrypted_secretkey = ssb.encrypt(&secretkey_vec, &encryption_key);
+            self.secret_key = secretKeyFromVec(encrypted_secretkey);
+            self.key_hash = key_hash;
+        }
+
+        self.write_to_file(path);
+        Ok(())
+    }
+
+    pub fn from_file(path: &Path) -> Self {  
         let mut file = File::open(path).expect("Could not open key file.");
         let mut content = String::new();
         file.read_to_string(&mut content).expect("Could not read from key file.");
         serde_json::from_str(&content).expect("Failed to deserialize KeyFile")
     }
+    
+    // decrypts a file and return an instance of the file
+    pub fn from_encrypted_file(path: &Path, key : &Key) -> Self {
+        let key_file = KeyFile::from_file(path);
+        let ssb = SodiumoxideSecretBox::new();
+        let decrypted_secretkey = ssb.decrypt(&secretKeyToVec(key_file.secret_key), &key);
+        KeyFile {
+            account_id: key_file.account_id,
+            public_key: key_file.public_key,
+            secret_key: secretKeyFromVec(decrypted_secretkey),
+            key_hash: KeyHash::default(),
+        }
+    }
+
+    // if a private key is encrypted , ask for a password in console
+    // if an entered password matches a password entered in step 3, decode an encrypted password otherwise keep asking for a correct password
+    pub fn from_encrypted_file_with_password(path : &Path) -> Result<Self, Error>  {  
+        let mut key_file = KeyFile::from_file(path);
+        if key_file.key_hash.eq(&KeyHash::default()) {
+            let password_input = PasswordInput::process()?; // ask for a password in command line
+            if  password_input.not_empty() { // password is not empty decode a file with a passwordm if password empty do nothing
+                let decryption_keyhash = KeyHash::new(&password_input.content);
+                if decryption_keyhash.eq(&key_file.key_hash) { // if hashes of passwords match then decrypt Keyfile.secret_key
+                    let decryption_key = password_input.gen_key_from_password_input()?;
+                    key_file = KeyFile::from_encrypted_file(path, &decryption_key);
+                }
+            }
+        }
+
+        Ok(key_file)
+    }
+}
+
+pub fn perform_encryption(
+    key_file : &mut KeyFile,
+    key_store_path: &Path,
+    home_dir: &Path,
+) -> Result<(), Error> {
+    key_file.write_to_file(&key_store_path);
+    
+    if home_dir.join("config.json").exists() {
+        //add password corrrectness validation to from_encrypted_file_with_password
+        KeyFile::from_encrypted_file_with_password(&key_store_path)?;
+    } else {
+        key_file.write_to_encrypted_file(&key_store_path)?;
+    }
+
+    Ok(())
+}
+
+pub struct PasswordInput {
+    pub content: Vec<u8>,
+    pub length: usize
```

Comment:
No need to store `length` separately. Just change the signature:
`pub fn enter_from_console() -> Resut<Self> {`
And do max length check inside.

Reply:
Then use `PasswordInput` as a simple wrapper around `Vec<u8>`, e.g.: `pub struct PasswordInput(Vec<u8>)`. However, I think there is no reason to not store password as `String`, since it comes as string from the console, e.g.`pub struct PasswordInput(String)`